### PR TITLE
read training strategy from config

### PR DESCRIPTION
**Pull request by katrina433**
Summary:

* Read training strategy from config file
* Add `strategy: ddp` to all existing config files
* Can also pass in strategy from the terminal by `python train.py config=configs/pretraining/debug.yaml training.lightning.strategy=ddp`
* Priority: terminal argument > config

Test plan:

logged `self.trainer.strategy` in `FLAVAPreTrainingLightningModule`

test 1
* command: `python train.py config=configs/pretraining/debug.yaml`
* config: `strategy: ddp`
* result: `DDPStrategy`

test 2
* command: `python train.py config=configs/pretraining/debug.yaml training.lightning.strategy=deepspeed`
* config: `strategy: ddp`
* result: `DeepSpeedStrategy`

### CODE REVIEW

**Review by ebsmothers — COMMENTED** [review]
The changes look good to me. Since it's hard to add a unit test to test these changes, it would be good to provide more detail in the test plan though. Can you share commands you ran and the resulting outputs for a few different cases? (e.g. changing the default value or passing a command-line arg and demonstrating it results in the expected value in the trainer)

**Comment by katrina433** [comment]
> The changes look good to me. Since it's hard to add a unit test to test these changes, it would be good to provide more detail in the test plan though. Can you share commands you ran and the resulting outputs for a few different cases? (e.g. changing the default value or passing a command-line arg and demonstrating it results in the expected value in the trainer)

updated description

File: examples/flava/train.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def main():
             LearningRateMonitor(logging_interval="step"),
             MultimodalEvalCallback(imagenet_datamodule=imagenet_datamodule),
         ],
-        strategy="ddp",
+        strategy=config.training.strategy,
```
Comment (ankitade) [inline examples/flava/train.py:62]: i dont think u need this. line 57 should take care of this

File: examples/flava/definitions.py [hunk 2]
```diff
@@ -74,6 +74,7 @@ class TrainingArguments:
     adam_weight_decay: float = 0.01
     adam_betas: Tuple[float, float] = field(default_factory=lambda: (0.9, 0.999))
     warmup_steps: int = 2000
+    strategy: str = "ddp"
```
Comment (ankitade) [inline examples/flava/definitions.py:77]: same as other comment. this should be specifiable through lightning config in line 68

**Review by ankitade — CHANGES_REQUESTED** [review]
thanks for the changes. after u try out my suggestions, can u paste the logs from lightning in the test plan instead of relying on prints.

**Comment by katrina433** [comment]
> thanks for the changes. after u try out my suggestions, can u paste the logs from lightning in the test plan instead of relying on prints.

Updated

File: examples/flava/utils.py [hunk 3]
```diff
@@ -38,4 +38,8 @@ def build_config():
         "max_steps" in config.training.lightning
     ), "lightning config must specify 'max_steps'"
 
+    assert (
```
Comment (ankitade) [inline examples/flava/utils.py:41]: do we need this? lightning works without any strategy override i thought

Comment (katrina433) [reply examples/flava/utils.py]: removed

**Review by ankitade — APPROVED** [review]

**Pull request closed without merging**